rolepermissions/shortcuts.py

```diff
@@ -6,91 +6,114 @@
+    return _assign_or_remove_role(user, role, "assign_role_to_user")
+
+
+def remove_role(user, role):
+    """Remove a role from a user."""
+    return _assign_or_remove_role(user, role, "remove_role_from_user")
+
+
+def clear_roles(user):
+    """Remove all roles from a user."""
+    roles = get_user_roles(user)
+
+    for role in roles:
+        role.remove_role_from_user(user)
+
+    return roles
 
 
 # Permissions
 
 
 def get_permission(permission_name):
+    """Get a Permission object from a permission name."""
     user_ct = ContentType.objects.get_for_model(get_user_model())
-    permission, created = Permission.objects.get_or_create(
-        content_type=user_ct,
-        codename=permission_name)
+    permission, _created = Permission.objects.get_or_create(content_type=user_ct, codename=permission_name)
 
     return permission
 
 
 def available_perm_status(user):
-    from rolepermissions.verifications import has_permission
-
-    role = get_user_role(user)
-    permission_names = role.permission_names_list()
-
+    """Get a boolean map of the permissions available to a user based on that user's roles."""
+    roles = get_user_roles(user)
     permission_hash = {}
 
-    for permission_name in permission_names:
-        permission_hash[permission_name] = has_permission(user, permission_name)
+    for role in roles:
+        permission_names = role.permission_names_list()
+
+        for permission_name in permission_names:
+            permission_hash[permission_name] = get_permission(permission_name) in user.user_permissions.all()
 
     return permission_hash
 
 
 def grant_permission(user, permission_name):
-    role = get_user_role(user)
+    """
+    Grant a user a specified permission.
+
+    Permissions are only granted if they are available in any of the user's roles.
+    """
+    roles = get_user_roles(user)
 
-    if role and permission_name in role.permission_names_list():
-        permission = get_permission(permission_name)
-        user.user_permissions.add(permission)
-        return True
+    for role in roles:
+        if permission_name in role.permission_names_list():
+            permission = get_permission(permission_name)
+            user.user_permissions.add(permission)
+            return True
 
     return False
 
 
 def revoke_permission(user, permission_name):
-    role = get_user_role(user)
-
-    if role and permission_name in role.permission_names_list():
-        permission = get_permission(permission_name)
-        user.user_permissions.remove(permission)
-        return True
+    """
+    Revoke a specified permission from a user.
+
+    Permissions are only revoked if they are available in any of the user's roles.
+    """
+    roles = get_user_roles(user)
+
+    for role in roles:
+        if permission_name in role.permission_names_list():
+            permission = get_permission(permission_name)
+            user.user_permissions.remove(permission)
+            return True
 
     return False
```

Comment:
What if a permission was added manually (outside of django-role-permissions) and the dev assumes this method will still work for that permission? I'd say also add the exception here.